src/utils/compile.rs

```diff
@@ -0,0 +1,233 @@
+use std::ffi::CString;
+use std::os::raw::c_void;
+
+/// compile a script, will result in a JSValueRef with tag JS_TAG_FUNCTION_BYTECODE or JS_TAG_MODULE.
+///  It can be executed with run_compiled_function().
+#[allow(dead_code)]
+pub fn compile<'a>(
+    context: &'a ContextWrapper,
+    script: &str,
+    file_name: &str,
+) -> Result<OwnedValueRef<'a>, ExecutionError> {
+    let filename_c = CString::new(file_name)
+        .map_err(|_e| ExecutionError::Internal("cstring creation failed".to_string()))?;
+    let code_c = CString::new(script)
+        .map_err(|_e| ExecutionError::Internal("cstring creation failed".to_string()))?;
+
+    let value_raw = unsafe {
+        q::JS_Eval(
+            context.context,
+            code_c.as_ptr(),
+            script.len() as _,
+            filename_c.as_ptr(),
+            q::JS_EVAL_FLAG_COMPILE_ONLY as i32,
+        )
+    };
+
+    // check for error
+    let ret = OwnedValueRef::new(context, value_raw);
+    if ret.is_exception() {
+        let ex_opt = context.get_exception();
+        if let Some(ex) = ex_opt {
+            Err(ex)
+        } else {
+            Err(ExecutionError::Internal("Unknown error".to_string()))
+        }
+    } else {
+        Ok(ret)
+    }
+}
+
+/// run a compiled function, see compile for an example
+#[allow(dead_code)]
+pub fn run_compiled_function<'a>(
+    context: &'a ContextWrapper,
+    compiled_func: &OwnedValueRef,
+) -> Result<OwnedValueRef<'a>, ExecutionError> {
+    assert!(compiled_func.is_compiled_function());
+    let val = unsafe { q::JS_EvalFunction(context.context, *compiled_func.as_inner_dup()) };
+    let val_ref = OwnedValueRef::new(context, val);
+    if val_ref.is_exception() {
+        let ex_opt = context.get_exception();
+        if let Some(ex) = ex_opt {
+            Err(ex)
+        } else {
+            Err(ExecutionError::Internal(
+                "run_compiled_function failed and could not get exception".to_string(),
+            ))
+        }
+    } else {
+        Ok(val_ref)
+    }
+}
+
+/// write a function to bytecode
+#[allow(dead_code)]
+pub fn to_bytecode(context: &ContextWrapper, compiled_func: &OwnedValueRef) -> Vec<u8> {
+    assert!(compiled_func.is_compiled_function());
+
+    let mut len = 0;
+
+    let slice_u8 = unsafe {
+        q::JS_WriteObject(
+            context.context,
+            &mut len,
+            *compiled_func.as_inner(),
+            q::JS_WRITE_OBJ_BYTECODE as i32,
+        )
+    };
+
+    let slice = unsafe { std::slice::from_raw_parts(slice_u8, len as usize) };
+    // it's a shame to copy the vec here but the alternative is to create a wrapping struct which free's the ptr on drop
+    let ret = slice.to_vec().clone();
+    unsafe { q::js_free(context.context, slice_u8 as *mut c_void) };
+    ret
+}
+
+/// read a compiled function from bytecode, see to_bytecode for an example
+#[allow(dead_code)]
+pub fn from_bytecode(
+    context: &ContextWrapper,
+    bytecode: Vec<u8>,
```

Comment:
```suggestion
    bytecode: &[u8],
```

Reply:
needed some extra code changes, added in separate commit